### PR TITLE
[microNPU] Add support for requantize

### PR DESCRIPTION
**Pull request by lhutton1**
Adds support for stand-alone requantize operation which is legalized to an identity operation on the NPU.

cc @manupa-arm @mbaret @ekalda @NicolaLancellotti @dchauhan-arm @jacobbohlin

### CODE REVIEW

**Review by jacobbohlin — APPROVED** [review]
LGTM

File: tests/python/contrib/test_ethosu/test_codegen.py [hunk 1]
```diff
@@ -949,7 +949,10 @@ def concat_func(*inputs):
         op = tf.concat(list(inputs), axis)
         return op
 
-    _compare_tvm_with_tflite(concat_func, shapes, accel_type)
+    # TODO(lhutton1) For now output is not bit exact with TFLite.
+    # This is because TFLite reference kernels are not being used.
+    # For this, TFLite will need upgrading to 2.6.
+    _compare_tvm_with_tflite(concat_func, shapes, accel_type, output_tolerance=1)
```
Comment (manupa-arm) [inline tests/python/contrib/test_ethosu/test_codegen.py:955]: Why is this changed now ?

File: tests/python/contrib/test_ethosu/test_legalize.py [hunk 2]
```diff
@@ -1502,5 +1503,61 @@ def verify(ext_func):
     verify(mod["tvmgen_default_ethos_u_main_0"])
 
 
+@pytest.mark.parametrize(
+    "ifm_shape,ifm_scale,ifm_zp,ofm_scale,ofm_zp",
+    [[(1, 8, 8, 3), 1.0, 0, 1.0, 0], [(1, 20, 30, 3), 1.345, 34, 0.32, -23]],
+)
+def test_ethosu_requantize(ifm_shape, ifm_scale, ifm_zp, ofm_scale, ofm_zp):
```
Comment (manupa-arm) [inline tests/python/contrib/test_ethosu/test_legalize.py:1510]: Since requantize is a sub-pattern of other patterns -- we would need a test to ensure that such offloads are not affected in the presense of extra requantize ops.

**Review by manupa-arm — CHANGES_REQUESTED** [review]
Mostly looks good!.
One question and a request to add a further test.

File: tests/python/contrib/test_ethosu/test_legalize.py (hunk 2 above)
Comment (lhutton1) [reply tests/python/contrib/test_ethosu/test_legalize.py:1510]: Done - I'm hoping I understood correctly, let me know if not!

File: tests/python/contrib/test_ethosu/test_codegen.py (hunk 1 above)
Comment (lhutton1) [reply tests/python/contrib/test_ethosu/test_codegen.py:956]: `qnn.concatenate` is lowered to a concatenate+requantize. Previously the requantize would not be offloaded, but now it is. There is a numerical inaccuracy in this case using the current version of TFLite, but we should see this issue when we upgrade to TFLite 2.6 and use the reference kernels that are made available.

Comment (ekalda) [reply tests/python/contrib/test_ethosu/test_codegen.py:956]: Thinking about it, I'm quite surprised that previously the concat on U55 + requantize on CPU yielded a bit exact result...

File: tests/python/contrib/test_ethosu/test_codegen.py [hunk 3]
```diff
@@ -987,5 +990,36 @@ def split_func(x):
...
+@pytest.mark.parametrize(
+    "ifm_shape,ifm_scale,ifm_zp,ofm_scale,ofm_zp",
+    [
+        [(1, 8, 8, 3), 1.0, 0, 1.0, 0],
+        [(1, 20, 30, 3), 1.345, 34, 0.32, -23],
+    ],
+)
+def test_ethosu_requantize(accel_type, ifm_shape, ifm_scale, ifm_zp, ofm_scale, ofm_zp):
+    dtype = "int8"
+    ifm_shape = [1, 8, 8, 3]
```
Comment (ekalda) [inline tests/python/contrib/test_ethosu/test_codegen.py:1003]: Looks like that `ifm_shape` always overwrites the `ifm_shape` that is specified by the parameter...

File: tests/python/contrib/test_ethosu/test_legalize.py [hunk 4]
```diff
@@ -1502,5 +1503,105 @@ def verify(ext_func):
...
+
+    mod["tvmgen_default_ethos_u_main_0"] = dataflow_pattern.rewrite(
+        rewriter, mod["tvmgen_default_ethos_u_main_0"]
+    )
+    verify(mod["tvmgen_default_ethos_u_main_0"])
+
+
+def test_multiple_requantize_offload():
+    """
+    Testing requantize offload in the case one requauntize operation is part of
```
Comment (ekalda) [inline tests/python/contrib/test_ethosu/test_legalize.py:1564]:
```suggestion
    Testing requantize offload in the case one requantize operation is part of
```

**Review by ekalda — COMMENTED** [review]
Some nits, but otherwise looks good! :)

File: tests/python/contrib/test_ethosu/test_codegen.py (hunk 3 above)
Comment (lhutton1) [reply tests/python/contrib/test_ethosu/test_codegen.py]: oops, thanks!

File: tests/python/contrib/test_ethosu/test_legalize.py (hunk 4 above)
Comment (lhutton1) [reply tests/python/contrib/test_ethosu/test_legalize.py]: good catch, thanks!

**Review by manupa-arm — APPROVED** [review]
LGTM!

**Review by ekalda — APPROVED** [review]
LGTM! :)

**Comment by lhutton1** [comment]
failure in `test_autotune` seems unrelated...
```
>           assert tuner.best_flops > 0
E           assert 0 > 0
E            +  where 0 = <tvm.autotvm.tuner.ga_tuner.GATuner object at 0x7fea7f3f2400>.best_flops
```
Retriggering.
